### DISCUSSION TITLE
FailedToExecuteCommand Exception on new MagickImageInfo("movie.AVI")

### DISCUSSION THREAD

**Discussion by ChuckSavage** [dlemstra/Magick.NET#731] [Help] [answered]
So, recently set my app to handle all Magick image format types. I get this exception on movie types, AVI, MP4 so far.

Here's the full exception:

$exception	{"FailedToExecuteCommand `\"ffmpeg.exe\" -nostdin -v -1 -i \"C:/Users/chuck/AppData/Local/Temp/magick-Ki7wx2uvxXyPx59zL2x-OE9JdWkkInzl\" -vframes 1 -vcodec pam -an -f rawvideo -y \"C:/Users/chuck/AppData/Local/Temp/magick-Gq2sUH_BgZlYDwrEzFebz_KmEmhsoax3.pam\"' (-1) @ error/delegate.c/ExternalDelegateCommand/478"}	ImageMagick.MagickDelegateErrorException

I'm not sure if this is a bug, or something I'm doing wrong. Any ideas?

**Comment by ChuckSavage** [comment]
Got a similar exception with PDF file type.

ImageMagick.MagickDelegateErrorException
  HResult=0x80131500
  Message=FailedToExecuteCommand `"gswin32c.exe" -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 "-sDEVICE=pngalpha" -dTextAlphaBits=4 -dGraphicsAlphaBits=4 "-r72x72" -dFirstPage=1 -dLastPage=1 "-sOutputFile=C:/Users/chuck/AppData/Local/Temp/magick-pNswIrJ0l-2Yy-ZutzMtkp84RPVUe0CX%d" "-fC:/Users/chuck/AppData/Local/Temp/magick-HX3Am-yUfFOFtRaFsJ7-1RTOnHkJX7Ft" "-fC:/Users/chuck/AppData/Local/Temp/magick-x_LsbR3Nfyw-wWmMtmP1hKzNfe9IWY6O"' (The system cannot find the file specified.
) @ error/delegate.c/ExternalDelegateCommand/475
  Source=Magick.NET-Q16-AnyCPU
  StackTrace:
   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result)
   at ImageMagick.MagickImage.NativeMagickImage.ReadFile(IMagickSettings`1 settings)
   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings, Boolean ping)
   at ImageMagick.MagickImageInfo.Read(String fileName)
   at ImageMagick.MagickImageInfo..ctor(String fileName)

**Comment by ChuckSavage** [comment]
MPG files give:

ImageMagick.MagickDelegateErrorException
  HResult=0x80131500
  Message=FailedToExecuteCommand `"ffmpeg.exe" -nostdin -v -1 -i "C:/Users/chuck/AppData/Local/Temp/magick-kfnKqR_LtKEFG8PGXbhG5fD4-bnNaQXP" -vframes 1 -vcodec pam -an -f rawvideo -y "C:/Users/chuck/AppData/Local/Temp/magick-lOUVGcQ5-McfuGT9I8M37Qq4w6Q8ddNK.pam"' (-1) @ error/delegate.c/ExternalDelegateCommand/478
  Source=Magick.NET-Q16-AnyCPU
  StackTrace:
   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result)
   at ImageMagick.MagickImage.NativeMagickImage.ReadFile(IMagickSettings`1 settings)
   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings, Boolean ping)
   at ImageMagick.MagickImageInfo.Read(String fileName)
   at ImageMagick.MagickImageInfo..ctor(String fileName)

**Comment by ChuckSavage** [comment]
MPEG files give:

ImageMagick.MagickDelegateErrorException
  HResult=0x80131500
  Message=FailedToExecuteCommand `"ffmpeg.exe" -nostdin -v -1 -i "C:/Users/chuck/AppData/Local/Temp/magick-wWOtc2F8EphWfM4RE1NNt5OD9ZVYwGp-" -vframes 1 -vcodec pam -an -f rawvideo -y "C:/Users/chuck/AppData/Local/Temp/magick--ArM68JntDKcYlJJbw7gxEylr857r3LV.pam"' (-1) @ error/delegate.c/ExternalDelegateCommand/478
  Source=Magick.NET-Q16-AnyCPU
  StackTrace:
   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result)
   at ImageMagick.MagickImage.NativeMagickImage.ReadFile(IMagickSettings`1 settings)
   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings, Boolean ping)
   at ImageMagick.MagickImageInfo.Read(String fileName)
   at ImageMagick.MagickImageInfo..ctor(String fileName)

**Comment by ChuckSavage** [comment]
MOV files give:

ImageMagick.MagickDelegateErrorException
  HResult=0x80131500
  Message=FailedToExecuteCommand `"ffmpeg.exe" -nostdin -v -1 -i "C:/Users/chuck/AppData/Local/Temp/magick-hdbOPFH1ySfKvW5zDTkK7f_Yo6bH-ugs" -vframes 1 -vcodec pam -an -f rawvideo -y "C:/Users/chuck/AppData/Local/Temp/magick-NeEDrpCsJDG_arKWgk1nXvpVaestLMVX.pam"' (-1) @ error/delegate.c/ExternalDelegateCommand/478
  Source=Magick.NET-Q16-AnyCPU
  StackTrace:
   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result)
   at ImageMagick.MagickImage.NativeMagickImage.ReadFile(IMagickSettings`1 settings)
   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings, Boolean ping)
   at ImageMagick.MagickImageInfo.Read(String fileName)
   at ImageMagick.MagickImageInfo..ctor(String fileName)

**Comment by ChuckSavage** [comment]
WMV gives:

ImageMagick.MagickDelegateErrorException
  HResult=0x80131500
  Message=FailedToExecuteCommand `"ffmpeg.exe" -nostdin -v -1 -i "C:/Users/chuck/AppData/Local/Temp/magick-iSgmJDidXqvbOqmZ-lu-YkiXkI29hY-C" -vframes 1 -vcodec pam -an -f rawvideo -y "C:/Users/chuck/AppData/Local/Temp/magick-QCbXQyz5SvOsPctbO2WPAmpzPgi-njAg.pam"' (-1) @ error/delegate.c/ExternalDelegateCommand/478
  Source=Magick.NET-Q16-AnyCPU
  StackTrace:
   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result)
   at ImageMagick.MagickImage.NativeMagickImage.ReadFile(IMagickSettings`1 settings)
   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings, Boolean ping)
   at ImageMagick.MagickImageInfo.Read(String fileName)
   at ImageMagick.MagickImageInfo..ctor(String fileName)

**Comment by dlemstra** [comment, marked as answer]
Some formats require an extra executable in the `bin` folder of your project. For movie type files you will need ffmpeg. If you want to read PDF files you will need Ghostscript installed on your machine.

**Reply by Yakonrus** [reply]
Hi.
Is it possible to somehow set the path to ffmpeg?
The fact is that a special assembly is used and it lies in a separate folder with additional files and a license.

**Reply by dlemstra** [reply]
ImageMagick expects `ffmpeg` to be inside the current folder or a folder specified in `%PATH%` so you could add your folder there.

**Reply by sanjay-vasudeva** [reply]
> Some formats require an extra executable in the `bin` folder of your project. For movie type files you will need ffmpeg. If you want to read PDF files you will need Ghostscript installed on your machine.

var image = MagickImage.FromBase64(req.Base64_string);

When I load an EPS format image like above, it throws below error ,
ImageMagick.MagickDelegateErrorException
HResult=0x80131500
Message=FailedToExecuteCommand "gswin32c.exe" .......

I want to check if an image format is readable or not like below before loading the image into imagemagick to avoid these errors,

MagickNET.SupportedFormats.Any(format => format.IsReadable &&
         format.Format.ToString().Equals(extension.Substring(1), StringComparison.OrdinalIgnoreCase))

Eventhough i dont have ghostscript dll in my folder, it says that Eps format is readable in the MagickFormatInfo. 
Is there any other way to check if i can read an image before loading the base64 into imagemagick?

**Reply by dlemstra** [reply]
The coder is reporting that it can read the format but that does not mean it can actually read the file. And because we use an external program you will only know if it works when you try it.

What you could do that is to add a tiny PDF file to your project and attempt to read this at the start of your application.

**Reply by sanjay-vasudeva** [reply]
Appreciate your swift response.
But this is quiet misleading for magick .NET library users right. Also the file format i will be receiving is not under my control. so I cannot check before hand. It would be really nice if the MagickNET.SupportedFormats gives the actual list of files that the library can read

**Reply by dlemstra** [reply]
For various image formats we use an external program to read the image. We don't want to check for the existence of those programs everytime we start the ImageMagick library.

**Reply by sanjay-vasudeva** [reply]
then instead of throwing MagickDelegateErrorException, if we have an exception common for all non-readable files like FormatNotSupported Exception with some info in it like "Ghostscript dll is missing" would be great. Because i am not sure which exception i will receive when i load the base64, which makes it impossible to handle exceptions properly.

**Reply by dlemstra** [reply]
We are unable to do that. When we execute the command of the PDF coder we are unable to determine if it fails because the PDF is invalid or because the Ghostscript executable could not be found.

Maybe the solution for your situation would be to disable the PDF coder?

**Comment by ChuckSavage** [comment]
Ah, okay - thanks.
